odl/space/weighting.py

```diff
@@ -0,0 +1,1047 @@
+        return self._impl
+
+    @property
+    def exponent(self):
+        """Exponent of this weighting."""
+        return self._exponent
+
+    def __eq__(self, other):
+        """Return ``self == other``.
+
+        Returns
+        -------
+        equal : `bool`
+            `True` if ``other`` is a the same weighting, `False`
+            otherwise.
+
+        Notes
+        -----
+        This operation must be computationally cheap, i.e. no large
+        arrays may be compared element-wise. That is the task of the
+        `equiv` method.
+        """
+        return (isinstance(other, WeightingBase) and
+                self.exponent == other.exponent and
+                self._dist_using_inner == other._dist_using_inner and
+                self.impl == other.impl)
+
+    def equiv(self, other):
+        """Test if ``other`` is an equivalent weighting.
+
+        Should be overwritten, default tests for equality.
+
+        Returns
+        -------
+        equivalent : `bool`
+            `True` if ``other`` is a `WeightingBase` instance which
+            yields the same result as this inner product for any
+            input, `False` otherwise.
+        """
+        return self == other
+
+    def inner(self, x1, x2):
+        """Calculate the inner product of two vectors.
+
+        Parameters
+        ----------
+        x1, x2 : `LinearSpaceVector`
+            Vectors whose inner product is calculated
+
+        Returns
+        -------
+        inner : `float` or `complex`
+            The inner product of the two provided vectors
+        """
+        raise NotImplementedError
+
+    def norm(self, x):
+        """Calculate the norm of a vector.
+
+        This is the standard implementation using `inner`.
+        Subclasses should override it for optimization purposes.
+
+        Parameters
+        ----------
+        x1 : `LinearSpaceVector`
+            Vector whose norm is calculated
+
+        Returns
+        -------
+        norm : `float`
+            The norm of the vector
+        """
+        return float(sqrt(self.inner(x, x).real))
+
+    def dist(self, x1, x2):
+        """Calculate the distance between two vectors.
+
+        This is the standard implementation using `norm`.
+        Subclasses should override it for optimization purposes.
+
+        Parameters
+        ----------
+        x1, x2 : `LinearSpaceVector`
+            Vectors whose mutual distance is calculated
+
+        Returns
+        -------
+        dist : `float`
+            The distance between the vectors
+        """
+        if self._dist_using_inner:
+            dist_squared = (self.norm(x1) ** 2 + self.norm(x2) ** 2 -
+                            2 * self.inner(x1, x2).real)
+            if dist_squared < 0:  # Compensate for numerical error
+                dist_squared = 0.0
+            return float(sqrt(dist_squared))
+        else:
+            return self.norm(x1 - x2)
+
+
+class MatrixWeighting(WeightingBase):
```

Comment:
It took me quite a while to figure out that this is not a implementation of `WeightingBase` but an abstract subclass, perhaps rename to `MatrixWeightingBase`?